text/0000-cargo-rustup-discovery.md

```diff
@@ -0,0 +1,457 @@
+
+It is not clear to me how this explicit trust system would interact with "safe directories".
+If it is added in the future, there may be some overlap or confusion on the difference between the two.
+
+## Granular capabilities
+
+An alternative is to associate actions and config options with granular capabilities.
+Then, Cargo and Rustup could trust files that are owned by different users if they don't
+present a hazard (such as executing external commands).
+This is the approach being taken by [gitoxide](https://github.com/Byron/gitoxide).
+It has a [`git-sec`](https://github.com/Byron/gitoxide/discussions/425) crate defining a security model.
+
+The Git developers also considered only generating an error if the ownership differed from the current user *and* there is a `core.fsmonitor` setting in the repo.
+However, they decided not to go that route.
+
+We feel like it is unlikely we can securely add such a mechanism to Cargo and Rustup due to the complexity.
+Also, in many cases, Cargo projects run code on the host
+(such as proc-macros and build scripts)
+which would severely limit what would be constituted as "safe".
+
+## Ceiling directories
+
+Git also has `GIT_CEILING_DIRECTORIES` for limiting the upwards search,
+and was presented as a way for git users to protect themselves.
+This can be used to avoid potentially loading malicious repositories,
+but in practice it is best used for improving performance.
+It might be nice for Cargo and Rustup to adopt something similar,
+but it is not a replacement for ownership checks since it is opt-in.
+
+## Honor the user of the current directory
+
+The ownership check could also include the owner of the current working directory as "approved" and allow any files owned by that user in any parent directory.
+This is under the assumption that the user running the command makes an explicit choice as to which directory they run Cargo or Rustup from.
+This could significantly ease the burden of dealing with this issue,
+as it will avoid needing to set safe directories in almost all use cases where they would be needed.
+
+However, I feel like this approach is too risky.
+For `git`, this would be risky because users often run `git` from the shell prompt.
+This means that just changing directory to another user's directory could expose them.
+This may also apply to Rust users.
+For example [Starship](https://starship.rs/) can execute `rustc` to get its version for the shell prompt.
+
+## Filesystem behavior risks
+
+This fix assumes that a user can't in any way create a file owned by another user
+(unless they have elevated permissions like root).
+I believe this is the case for all major operating systems and filesystems.
+
+One (major) known exception is if a network-mounted filesystem is configured to map all files to the local user, which is not uncommon.
```

Comment:
Another major exception is UID reuse – where the attacker creates an user, poisons the filesystem and then later a genuine user gets created at that specific UID.

systemd comes to mind as an example of a tool that a) creates users left and right; and b) has mitigations for this problem.

Reply:
Thanks! I have added that as another concern. I personally don't feel like the risk here is very high, as there are likely other avenues than Rustup/Cargo for escalating once an attacker has control of a temporary new user. However, it definitely is a concern.